src/lib/routers/history.ts

```diff
@@ -0,0 +1,200 @@
+  const queryString = qsModule.stringify(routeState);
+  const portWithPrefix = port === '' ? '' : `:${port}`;
+
+  // IE <= 11 has no proper `location.origin` so we cannot rely on it.
+  if (!routeState || Object.keys(routeState).length === 0) {
+    return `${protocol}//${hostname}${portWithPrefix}${pathname}${hash}`;
+  }
+
+  return `${protocol}//${hostname}${portWithPrefix}${pathname}?${queryString}${hash}`;
+};
+
+const defaultParseURL: BrowserHistoryProps['parseURL'] = ({
+  qsModule,
+  location,
+}): RouteState => {
+  // `qs` by default converts arrays with more than 20 items to an object.
+  // We want to avoid this because the data structure manipulated can therefore vary.
+  // Setting the limit to `100` seems a good number because the engine's default is 100
+  // (it can go up to 1000 but it is very unlikely to select more than 100 items in the UI).
+  //
+  // Using an `arrayLimit` of `n` allows `n + 1` items.
+  //
+  // See:
+  //   - https://github.com/ljharb/qs#parsing-arrays
+  //   - https://www.algolia.com/doc/api-reference/api-parameters/maxValuesPerFacet/
+  return qsModule.parse(location.search.slice(1), { arrayLimit: 99 });
+};
+
+const setWindowTitle = (title?: string): void => {
+  if (title) {
+    window.document.title = title;
+  }
+};
+
+class BrowserHistory {
+  /**
+   * Transforms a UI state into a title for the page.
+   */
+  private readonly windowTitle?: BrowserHistoryProps['windowTitle'];
+  /**
+   * Time in milliseconds before performing a write in the history.
+   * It prevents from adding too many entries in the history and
+   * makes the back button more usable.
+   *
+   * @default 400
+   */
+  private readonly writeDelay: BrowserHistoryProps['writeDelay'];
+  /**
+   * Creates a full URL based on the route state.
+   * The storage adaptor maps all syncable keys to the query string of the URL.
+   */
+  private readonly _createURL: BrowserHistoryProps['createURL'];
+  /**
+   * Parses the URL into a route state.
+   * It should be symetrical to `createURL`.
+   */
+  private readonly parseURL: BrowserHistoryProps['parseURL'];
+
+  private writeTimer?: number;
+  private _onPopState?(event: PopStateEvent): void;
+
+  /**
+   * Initializes a new storage provider that syncs the search state to the URL
+   * using web APIs (`window.location.pushState` and `onpopstate` event).
+   */
+  public constructor(
+    {
+      windowTitle,
+      writeDelay = 400,
+      createURL = defaultCreateURL,
+      parseURL = defaultParseURL,
+    }: BrowserHistoryProps = {} as BrowserHistoryProps
+  ) {
+    this.windowTitle = windowTitle;
+    this.writeTimer = undefined;
+    this.writeDelay = writeDelay;
+    this._createURL = createURL;
+    this.parseURL = parseURL;
+
+    const title = this.windowTitle && this.windowTitle(this.read());
+
+    setWindowTitle(title);
+  }
+
+  /**
+   * Reads the URL and returns a syncable UI search state.
+   */
+  private read(): RouteState {
```

Comment:
read/write are also public, we call them from the `RoutingManager`.

https://github.com/algolia/instantsearch.js/blob/21e4d81b3e37c03686acaf3a8dc2c73c3c519570/src/lib/RoutingManager.js#L78

https://github.com/algolia/instantsearch.js/blob/21e4d81b3e37c03686acaf3a8dc2c73c3c519570/src/lib/RoutingManager.js#L12